metrics_sdk/lib/opentelemetry/sdk/metrics/state/metric_stream.rb

```diff
@@ -30,30 +30,60 @@ def initialize(
             @instrument_kind = instrument_kind
             @meter_provider = meter_provider
             @instrumentation_scope = instrumentation_scope
-            @aggregation = aggregation
+            @default_aggregation = aggregation
+            @data_points = {}
 
             @mutex = Mutex.new
           end
 
           def collect(start_time, end_time)
-            @mutex.synchronize do
-              MetricData.new(
-                @name,
-                @description,
-                @unit,
-                @instrument_kind,
-                @meter_provider.resource,
-                @instrumentation_scope,
-                @aggregation.collect(start_time, end_time),
-                @aggregation.aggregation_temporality,
-                start_time,
-                end_time
-              )
+            metric_data = []
+            registered_views = find_registered_view
+
+            if registered_views.empty?
+              metric_data << aggregate_metric_data(start_time, end_time)
+            else
+              registered_views.each { |view| metric_data << aggregate_metric_data(start_time, end_time, aggregation: view.aggregation) }
             end
+
+            metric_data
           end
 
           def update(value, attributes)
-            @mutex.synchronize { @aggregation.update(value, attributes) }
+            registered_views = find_registered_view
+            if registered_views.empty?
+              @mutex.synchronize { @default_aggregation.update(value, attributes, @data_points) }
+            else
+              registered_views.each do |view|
+                @mutex.synchronize do
+                  attributes ||= {}
+                  attributes.merge!(view.attribute_keys)
+                  view.aggregation.update(value, attributes, @data_points)
+                end
+              end
+            end
+          end
+
+          def aggregate_metric_data(start_time, end_time, aggregation: nil)
+            aggregator = aggregation || @default_aggregation
+            MetricData.new(
+              @name,
+              @description,
+              @unit,
+              @instrument_kind,
+              @meter_provider.resource,
+              @instrumentation_scope,
+              aggregator.collect(start_time, end_time, @data_points),
+              aggregator.aggregation_temporality,
+              start_time,
+              end_time
+            )
+          end
+
+          def find_registered_view
+            registered_views = []
+            @meter_provider.registered_views.each { |view| registered_views << view if view.match_instrument?(self) }
+            registered_views
           end
```

Comment:
I'm concerned about running `find_registered_view` every time we run the `collect` or `update` methods.

It seems like we may be duplicating work that could be preserved across runs.

Do we expect the registered views to change over time? Could we save the registered views we know match with a particular metrics stream? Or maybe just check to see if there are new views and add those to an existing list each time?

Reply:
I didn't see any place indicate user can or should add_view after create the instrument. Can we make that assumption? However, if that's the case, the instrument initialization order has to be
```
OpenTelemetry.meter_provider.add_view(...)
OpenTelemetry.meter_provider.add_view(...)
... # add more views

counter = meter.create_counter('counter') # instrument initialization
```
I updated the logic based on the assumption that registered views doesn't change for initialized instrument

Reply:
The initialization order is a good point. I'm not sure that's quite what we want. I think we should defer this point to another PR. I'll open a new issue and we can continue the conversation there.

Reply:
@xuan-cao-swi  - Here's the new issue to think about this further: https://github.com/open-telemetry/opentelemetry-ruby/issues/1672